src/requiredBy.js

```diff
@@ -0,0 +1,12 @@
+export default function requiredBy(requiredByPropName, propType, defaultValue = null) {
+  return function validator(props, propName, componentName, ...rest) {
+    if (props[requiredByPropName]) {
+      if (props[propName] === defaultValue || typeof props[propName] === 'undefined') {
+        return new TypeError(
+          `${componentName}: when ${requiredByPropName} is true, prop “${propName}” must be present.`,
+        );
+      }
+    }
+    return propType(props, propName, componentName, ...rest);
+  };
```

Comment:
isn't the same `required` logic being essentially checked here? https://github.com/airbnb/prop-types/pull/30/files#diff-df688d046a57d57b4df2a8a950118e1fR4
this way, it could even check against a optional passed in default value, since the default value from `defaultProps` is applied before going into the validator. (this is useful since we normally default props that are `PropType.func` to a no-op function instead of null.)